src/OptimizationDIExt.jl

```diff
@@ -150,12 +151,13 @@ function instantiate_function(
     end
 
     if f.cons_vjp === nothing && cons_vjp == true && cons !== nothing
-        extras_pullback = prepare_pullback(cons_oop, adtype, x)
-        function cons_vjp!(J, θ, v)
+        dy = similar(x)
+        extras_pullback = prepare_pullback(cons_oop, adtype, x, dy)
+        function cons_vjp!(J, v, θ)
             pullback!(cons_oop, J, adtype, θ, v, extras_pullback)
         end
     elseif cons_vjp == true && cons !== nothing
-        cons_vjp! = (J, θ, v) -> f.cons_vjp(J, θ, v, p)
+        cons_vjp! = (J, v, θ) -> f.cons_vjp(J, v, θ, p)
```

Comment:
Actually, I don't think this is correct. We don't want to have this order as API too since it'll be inconsistent with `jvp` and be higher mental load.

What makes you think the current implementation is incorrect? https://gdalle.github.io/DifferentiationInterface.jl/DifferentiationInterface/dev/api/#DifferentiationInterface.pullback and https://github.com/jump-dev/MathOptInterface.jl/blob/cb1ad41c3a5ce1d12e83f2be60f10b191b0e22c5/src/nlp.jl#L762-L809 both match my understanding as well 

I did spend a bit of time making sure the dimensions here were correct

Reply:
When you call DI, the argument order is `pullback!(cons_oop, J, adtype, θ, v, extras_pullback)`, but in SciML, the argument order is 
```julia
- `cons_jvp(Jv,v,x,p)` or `Jv=cons_jvp(v,x,p)`: the Jacobian-vector product of the constraints.
- `cons_vjp(Jv,v,x,p)` or `Jv=cons_vjp(v,x,p)`: the Jacobian-vector product of the constraints.
```
https://github.com/SciML/SciMLBase.jl/blob/master/src/scimlfunctions.jl#L1833C1-L1834C96

Reply:
Ah lol so I missed the reference closest to home, I'll change that, this order makes less sense to me than the one used here. What do you think?

Reply:
I don't have a strong opinion here. I'd almost expect `v'J(x)` to take args `v, x` and vice versa for `J(x)*v`, but having them be different might also be confusing. As long as SciML is internally consistent everywhere it can be whatever you prefer.

Reply:
makes sense, thanks for pointing these out 🙌